gtest/src/mailbox.rs

```diff
@@ -0,0 +1,99 @@
+use crate::{manager::ExtManager, CoreLog, Log, RunResult};
+use codec::Encode;
+use gear_core::message::{Dispatch, DispatchKind, StoredMessage};
+use gear_core::{
+    ids::{MessageId, ProgramId},
+    message::Payload,
+};
+use std::cell::RefCell;
+
+pub struct Mailbox<'a> {
+    manager_reference: &'a RefCell<ExtManager>,
+    program_id: ProgramId,
+}
+
+impl<'a> Mailbox<'a> {
+    pub(crate) fn new(
+        program_id: ProgramId,
+        manager_reference: &'a RefCell<ExtManager>,
+    ) -> Mailbox<'a> {
+        Mailbox {
+            program_id,
+            manager_reference,
+        }
+    }
+
+    pub fn contains(&self, log: Log) -> bool {
+        self.manager_reference
+            .borrow_mut()
+            .actor_to_mailbox
+            .get(&self.program_id)
+            .expect("No mailbox with such program id")
+            .iter()
+            .any(|message| log.eq(message))
+    }
+
+    pub fn take_message(&self, log: Log) -> MessageReplier {
+        let index = self
+            .manager_reference
+            .borrow_mut()
+            .actor_to_mailbox
+            .get(&self.program_id)
+            .expect("No mailbox with such program id")
+            .iter()
+            .position(|message| log.eq(message))
+            .expect("No message that satisfies log");
+
+        let taken_message = self
+            .manager_reference
+            .borrow_mut()
+            .take_message(&self.program_id, index);
+
+        MessageReplier::new(taken_message, self.manager_reference)
+    }
+
+    pub fn reply(&self, log: Log, payload: impl Encode, value: u128) -> RunResult {
+        self.take_message(log).reply(payload, value)
+    }
+
+    pub fn reply_bytes(&self, log: Log, raw_payload: impl AsRef<[u8]>, value: u128) -> RunResult {
+        self.take_message(log).reply_bytes(raw_payload, value)
+    }
+}
+
+pub struct MessageReplier<'a> {
+    log: CoreLog,
+    manager_reference: &'a RefCell<ExtManager>,
+}
+
+impl<'a> MessageReplier<'a> {
```

Comment:
Why do we need a separate `MessageReplier` instead of reply methods within the `Mailbox`?

Reply:
It was @breathx idea. As for me that looks overweight, but Dima really liked this possibility:
```
message = mailbox.take_message(log);
result = message.reply(payload);
```

Reply:
I prefer to let more abilities to write test code for users. So I accept both variants: reply on message from mailbox or from got from mailbox message (non Copy, non Clone).

```rust
struct Mailbox<'a> {
    reff: &'a Manager,
    id: u8, // user id
}

// no Copy, no Clone
struct MailboxedMessage<'a> {
    log: CoreLog,
    reff: &'a Manager,
}

impl <'a> MailboxedMessage<'a> {
    pub fn reply(self, v: impl Encode) -> RunResult {
        self.reply_bytes(v.encode())
    }

    pub fn reply_bytes(self, v: impl AsRef<[u8]>) -> RunResult ;
}

impl<'a> Mailbox<'a> {
    pub fn contains(&self, log: Log) -> bool { false }

    pub fn take_msg(&self, log: Log) -> MailboxedMessage {

    }

    pub fn reply(&self, log: Log, v:  impl Encode) -> RunResult  {
        self.take_msg(log).reply(v)
    }

    pub fn reply_bytes(&self, log: Log, v: impl AsRef<[u8]>) -> RunResult {
        self.take_msg(log).reply_bytes(v)
    }
}

#[test]
fn name() {
    let sys = System::new();
    let prog = Program::current(&sys);

    let log = Log::builder().from(1).to(10).payload_bytes("pong");

    // user id = 10
    let res = prog.send_bytes(10, "ping");
    assert!(res.contains(&log));

    let mb: Mailbox = sys.get_mailbox(10);
    assert!(mb.contains(&log));

    let log_on_reply = Log::builder().from(1).to(10).payload_bytes("reply pong");

   // #1
    let res2 = mb.reply_bytes(&log, "reply ping");
    assert!(!res2.contains(& log_on_reply));

    // #2

    let mboxed_msg = mb.take_msg(&log);
    let res3 = mboxed_msg.reply_bytes(&log, "reply ping");
    assert!(res3.contains(& log_on_reply));
}
```

This code is just scheme*

Reply:
Agree, it's more flexible.